lib/gruf/controllers/request.rb

```diff
@@ -15,51 +15,75 @@
 # OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #
 module Gruf
-  module Hooks
+  module Controllers
     ##
-    # Base class for a hook. Define before, around, outer_around, or after methods to utilize functionality.
+    # Encapsulates a request for a controller
     #
-    class Base
-      include Gruf::Loggable
+    class Request
+      # @var [Object] message
+      attr_reader :message
+      # @var [GRPC::ActiveCall] active_call
+      attr_reader :active_call
+      # @var [Symbol] method_key
+      attr_reader :method_key
 
-      # @return [Gruf::Service] service The service to perform the hook against
-      attr_reader :service
-      # @return [Hash] options Options to use for the hook
-      attr_reader :options
+      delegate :metadata, to: :active_call
 
-      ##
-      # Initialize the hook and run setup
-      #
-      # @param [Gruf::Service] service The gruf service that the hook will perform against
-      # @param [Hash] options (Optional) A hash of options for this hook
-      #
-      def initialize(service, options = {})
+      def initialize(method_key:, service:, rpc_desc:, active_call:, message:)
+        @method_key = method_key
         @service = service
-        @options = options
-        setup
+        @rpc_desc = rpc_desc
+        @active_call = active_call
+        @message = message
       end
 
       ##
-      # Method that can be used to setup the hook prior to running it
+      # @return [Symbol]
       #
-      def setup
-        # noop
+      def type
+        case
+          when @rpc_desc.client_streamer?
+            :client_streamer
+          when @rpc_desc.server_streamer?
+            :server_streamer
+          when @rpc_desc.bidi_streamer?
+            :bidi_streamer
+          else
+            :request_response
+        end
       end
 
       ##
       # @return [String] Returns the service name as a translated name separated by periods
       #
       def service_key
-        service.class.name.underscore.tr('/', '.')
+        @service.name.underscore.tr('/', '.')
       end
 
       ##
       # Parse the method signature into a service.method name format
       #
       # @return [String] The parsed service method name
       #
-      def method_key(call_signature)
-        "#{service_key}.#{call_signature.to_s.gsub('_without_intercept', '')}"
+      def method_name
+        "#{service_key}.#{method_key}"
+      end
+
+      ##
+      # Return all messages for this request, properly handling different request types
+      #
+      # @return Enumerable<Object>
+      #
+      def messages
+        if type == :client_streamer
```

Comment:
🍺 IMHO, this class is asking for a `RequestType` object to manage the logic of working with a specific request type. That could help this class offload the task conditional around L40 in `type` by building the `RequestType` object from the `rpc_desc` and also delegate the formatting of the messages to the RequestType class instead of doing that here.

I understand this may be considered as either an SRP violation or a premature abstraction so that's why I'm making this suggestion as optional.

If you need any clarification on this suggestion just let me know.